### PR TITLE
Deploy to ECS

### LINKED ISSUE #192: Setup staging site with Continuous Deployment
**Describe the task**
We need to set up staging sites. The sites to be run at:
- Plio frontend: staging-app.plio.in
- Plio backend: staging-backend.plio.in

**Expected behavior**
1. Staging site fully functional
2. Continuous Deployment in place with GitHub action, ECR and ECS
3. Deployment and other relevant documentation updates

### PR DESCRIPTION
**Pull request by rathorevaibhav**
Fixes #192 

## Summary

1. Updates to dockerfile for build-time arguments
2. Github workflow to build image and push to ECR

## Test Plan

Not needed

### CODE REVIEW

File: docs/DEPLOYMENT.md [hunk 1]
```diff
@@ -2,26 +2,87 @@
...
+11. AWS Relational Database Service (RDS)
 
-#### Deployment steps
+#### Staging
+Setting up staging environment on AWS is pretty straightforward.
+1. Login to your AWS console.
+2. Go to VPC. (skip this step if you've already created a VPC when setting up backend repository)
+   1. Create a new VPC.
+   2. Name it `plio-staging-vpc`.
+   3. In IPv4 CIDR block, enter `10.0.0.0/28`.
```
[the remark this answers is not in the document]
Comment (dalmia) [reply docs/DEPLOYMENT.md:28]: default value - some link

File: docs/DEPLOYMENT.md [hunk 2]
```diff
@@ -2,26 +2,87 @@
...
+   4. Click on create button. You will see the new VPC under the list of VPCs.
+3. Create a new Elastic IP by going to EC2 dashboard and navigating to Elastic IP section.
```
Comment (dalmia) [inline docs/DEPLOYMENT.md:30]: used to give load balancer a permanent Ip address

File: docs/DEPLOYMENT.md [hunk 3]
```diff
@@ -2,26 +2,87 @@
...
+   1. Click on Allocate Elastic IP address and click on Allocate button.
+   2. You will see a new IP address in the IPs list. Name it `plio-frontend-staging`.
+4. Go to Target groups.
+   1. Create a new target group.
+   2. Choose target type to be `IP addresses`.
+   3. Name the target group as `plio-frontend-staging`.
```
[the remark this answers is not in the document]
Comment (dalmia) [reply docs/DEPLOYMENT.md:36]: yes

File: docs/DEPLOYMENT.md [hunk 4]
```diff
@@ -2,26 +2,87 @@
...
+7. Now go to ECS and create a new task definition with name `plio-frontend-staging`.
+   1. Set the task role as `ecsTaskExecutionRole`.
+   2. Set the task memory and task CPU based on your needs.
+   3. Create a new container with name `plio-frontend-staging`.
+   4. Enter port `80` in the port mapping field.
+   5. Save the container definition and the task definition.
+   6. You will see the new task definition within the list.
+8. Go to clusters and create a new cluster with name `plio-staging-cluster`. (skip this step if you've already created a VPC when setting up backend repository)
+   1. Use `Networking only` option
+   2. Create a new VPC for your cluster.
```
[the remark this answers is not in the document]
Comment (dalmia) [reply docs/DEPLOYMENT.md:57]: TODO: link to the older VPC

File: docs/DEPLOYMENT.md [hunk 5]
```diff
@@ -2,26 +2,87 @@
...
+   3. Click on create button.
+   4. You will see the new cluster within the list of clusters.
+9.  Get into `plio-staging-cluster` and create a new service.
+   5. Set launch type to Fargate. We'll use serverless deployments for Plio.
+   6. Name the service as `plio-staging-frontend`.
+   7. Under task definition, select `plio-frontend-staging` and use latest revision.
+   8. Number of tasks to be one.
+   9. Service type to be `REPLICA`.
+   10. Minimum healthy percentage should be 100 and maximum percent to be 200.
```
Comment (dalmia) [inline docs/DEPLOYMENT.md:66]: attempting to write 1-2 liners about the importance of this.

File: docs/DEPLOYMENT.md [hunk 6]
```diff
@@ -2,26 +2,87 @@
...
+   11. Deployment type to be `rolling update`.
+   12. Keep other values as default.
+   13. Use the Cluster VPC and the subnet that you configured with Elastic IP.
+   14. Auto-assign public IP to have `ENABLED`.
```
Comment (dalmia) [inline docs/DEPLOYMENT.md:70]: one line: else makes task not publicly accessible.

File: docs/DEPLOYMENT.md [hunk 7]
```diff
@@ -2,26 +2,87 @@
...
+   2. You will see a new IP address in the IPs list. Name it `plio-frontend-staging`.
+4. Go to Target groups.
+   1. Create a new target group.
+   2. Choose target type to be `IP addresses`.
+   3. Name the target group as `plio-frontend-staging`.
+   4. Select the `plio-staging-vpc` for the target group VPC.
+   5. Proceed to create target group. You will see target group in the list.
+5. Go to Load Balancers.
+   1. Create a new load balancer.
+   2. Select Network Load Balancer option.
```
Comment (dalmia) [inline docs/DEPLOYMENT.md:41]: one comment - to enable web socket connections which are not allowed in ALB

File: docs/DEPLOYMENT.md [hunk 8]
```diff
@@ -2,26 +2,87 @@
...
+10. AWS Identity and Access Management (IAM)
+11. AWS Relational Database Service (RDS)
 
-#### Deployment steps
+#### Staging
+Setting up staging environment on AWS is pretty straightforward.
+1. Login to your AWS console.
+2. Go to VPC. (skip this step if you've already created a VPC when setting up backend repository)
+   1. Create a new VPC.
+   2. Name it `plio-staging-vpc`.
```
[the remark this answers is not in the document]
Comment (rathorevaibhav) [reply docs/DEPLOYMENT.md:27]: Yeah I'll work on that.

File: .github/workflows/deploy_to_ecs_staging.yml [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        --build-arg VUE_APP_FRONTEND=$VUE_APP_FRONTEND \
+        --build-arg VUE_APP_I18N_LOCALE=$VUE_APP_I18N_LOCALE \
+        --build-arg VUE_APP_I18N_FALLBACK_LOCALE=$VUE_APP_I18N_FALLBACK_LOCALE \
+        --build-arg VUE_APP_GOOGLE_CLIENT_ID=$VUE_APP_GOOGLE_CLIENT_ID \
+        --build-arg VUE_APP_BACKEND_API_CLIENT_ID=$VUE_APP_BACKEND_API_CLIENT_ID \
+        --build-arg VUE_APP_BACKEND_API_CLIENT_SECRET=$VUE_APP_BACKEND_API_CLIENT_SECRET \
+        -t $ECR_REGISTRY/$ECR_REPOSITORY:$IMAGE_TAG .
+        echo "Pushing image to ECR..."
+        docker push $ECR_REGISTRY/$ECR_REPOSITORY:$IMAGE_TAG
+        echo "::set-output name=image::$ECR_REGISTRY/$ECR_REPOSITORY:$IMAGE_TAG"
```
[the remark this answers is not in the document]
Comment (rathorevaibhav) [reply .github/workflows/deploy_to_ecs_staging.yml]: Yep.

File: docs/DEPLOYMENT.md [hunk 10]
```diff
@@ -2,26 +2,95 @@
...
+   5. Check out this [AWS guide for more details on VPC](https://docs.aws.amazon.com/vpc/latest/userguide/working-with-vpcs.html).
+3. Create a new Elastic IP by going to EC2 dashboard and navigating to Elastic IP section.
+   1. Click on Allocate Elastic IP address and click on Allocate button.
+   2. You will see a new IP address in the IPs list. Name it `plio-frontend-staging`.
+   3. This will be used in later steps to give the load balancer a permanent IP address.
+4. Go to Target groups.
+   1. Create a new target group.
+   2. Choose target type to be `IP addresses`.
+   3. Name the target group as `plio-frontend-staging`.
+   4. Select the `plio-staging-vpc` for the target group VPC.
```
Comment (dalmia) [inline docs/DEPLOYMENT.md:39]: Had to assign an IP as a target here.

Comment (dalmia) [inline docs/DEPLOYMENT.md:39]: had to select Protocol as TCP instead of HTTP

File: docs/DEPLOYMENT.md [hunk 11]
```diff
@@ -2,26 +2,95 @@
...
+5. Go to Load Balancers.
+   1. Create a new load balancer.
+   2. Select Network Load Balancer option. We use NLB for easy support of web socket connections.
+   3. Name the LB as `plio-frontend-staging`.
+   4. Select the `plio-staging-vpc` for the load balancer.
+   5. In the subnet mappings, check the first desired zone and use Elastic IP under IPv4 settings for that subnet.
+   6. Under listeners and routing, select the target group `plio-frontend-staging` for TCP port 80.
+   7. Proceed to create load balancer. You will see load balancer in the list.
+6. Go to ECR and create a new repository named `plio-frontend-staging`.
+7. Now go to ECS and create a new task definition with name `plio-frontend-staging`.
```
Comment (dalmia) [inline docs/DEPLOYMENT.md:50]:
```suggestion
7. Now go to ECS and create a new task definition, select Fargate and use the name `plio-frontend-staging`.
```

File: docs/DEPLOYMENT.md [hunk 12]
```diff
@@ -2,26 +2,95 @@
...
+   1. Set the task role as `ecsTaskExecutionRole`.
+   2. Set the task memory and task CPU based on your needs.
+   3. Create a new container with name `plio-frontend-staging`.
+   4. Enter port `80` in the port mapping field.
```
Comment (dalmia) [inline docs/DEPLOYMENT.md:54]: need to add an image url

File: docs/DEPLOYMENT.md [hunk 13]
```diff
@@ -2,26 +2,95 @@
...
+   5. Save the container definition and the task definition.
+   6. You will see the new task definition within the list.
+8. Under ECS, go to clusters and create a new cluster with name `plio-staging-cluster`. (skip this step if you've already created a VPC when setting up backend repository)
+   1. Use `Networking only` option. We will go with serverless deployment so that we don't worry about managing our own server instances.
+   2. Don't create a new VPC for your cluster. We'll use the VPC created in previous step in the next step of creating a service.
+   3. Click on create button.
+   4. You will see the new cluster within the list of clusters.
+9.  Get into `plio-staging-cluster` and create a new service.
+   1. Set launch type to Fargate. We'll use serverless deployments for Plio.
+   2. Name the service as `plio-staging-frontend`.
```
Comment (dalmia) [inline docs/DEPLOYMENT.md:64]:
```suggestion
   2. Name the service as `plio-frontend-staging`.
```

**Review by dalmia — APPROVED** [review]

**Pull request merged**